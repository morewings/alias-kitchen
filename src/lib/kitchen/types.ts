export type InitialPaths = Record<string, string[]> | undefined;
