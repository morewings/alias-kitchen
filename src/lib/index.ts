export {kitchen} from './kitchen';
export {kitchen as default} from './kitchen';
