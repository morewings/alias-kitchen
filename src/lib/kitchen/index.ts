export {kitchen} from './kitchen.ts';
