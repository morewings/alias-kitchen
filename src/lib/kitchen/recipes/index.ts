export {babel} from './babel.ts';
export {jestRecipe} from './jest.ts';
export {rollup} from './rollup.ts';
export {webpack} from './webpack.ts';
