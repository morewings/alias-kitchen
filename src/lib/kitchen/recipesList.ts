export enum Recipes {
    rollup = 'rollup',
    webpack = 'webpack',
    vite = 'vite',
    jest = 'jest',
    babel = 'babel',
    rspack = 'rspack',
}
