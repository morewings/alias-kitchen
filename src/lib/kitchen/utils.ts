import type {InitialPaths} from '@/lib/kitchen/types.ts';

export const createPathArray = (initialString: string) => {
    return initialString.split('/');
};

export const createPathString = (initialPath: string[]) => {
    return initialPath.join('/');
};

export const removeLastSegment = (pathString: string, segmentName?: string) => {
    const path = createPathArray(pathString);
    const lastSegment = path[path.length - 1];
    if (lastSegment === segmentName || segmentName === undefined) {
        return createPathString(path.slice(0, path.length - 1));
    }
    return pathString;
};

export const removeFirstSegment = (pathString: string, segmentName?: string) => {
    const path = createPathArray(pathString);
    const firstSegment = path[0];
    if (firstSegment === segmentName || segmentName === undefined) {
        return createPathString(path.slice(1, path.length));
    }
    return pathString;
};

export const normalizePath = (initialPaths: InitialPaths) => {
    return Object.entries(initialPaths || {})
        .map(([alias, [directory]]) => [
            removeLastSegment(alias, '*'),
            removeLastSegment(directory, '*'),
        ])
        .map(([alias, directory]) => [removeLastSegment(alias, '/'), directory]);
};
