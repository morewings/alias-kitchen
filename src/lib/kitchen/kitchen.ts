import {getTsconfig} from 'get-tsconfig';

import type {InitialPaths} from './types.ts';
import {Recipes} from './recipesList.ts';
import {rollup} from './recipes/rollup.ts';
import {jestRecipe} from './recipes/jest.ts';
import {babel} from './recipes/babel.ts';
import {webpack} from './recipes/webpack.ts';

const recipesMapping = {
    [Recipes.rollup]: rollup,
    [Recipes.webpack]: webpack,
    [Recipes.vite]: rollup,
    [Recipes.jest]: jestRecipe,
    [Recipes.babel]: babel,
    [Recipes.rspack]: webpack,
};

export type Config = {
    recipe: keyof typeof Recipes;
};

export const kitchen = ({recipe}: Config) => {
    const initialPaths: InitialPaths = getTsconfig()?.config?.compilerOptions?.paths;
    return recipesMapping[recipe](initialPaths);
};
