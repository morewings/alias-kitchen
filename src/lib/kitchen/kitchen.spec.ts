import {kitchen} from '@/lib/kitchen';
import {Recipes} from '@/lib/kitchen/recipesList.ts';

jest.mock('get-tsconfig', () => ({
    __esModule: true,
    getTsconfig: () => ({
        config: {
            compilerOptions: {
                paths: {'@/*': ['./src/*']},
            },
        },
    }),
}));

jest.mock('path', () => ({
    __esModule: true,
    resolve: (directory: string) => `/resolved/${directory}`,
}));

const recipes = Object.values(Recipes);

describe('alias kitchen', () => {
    it.each(recipes)('cooks %s configuration', recipe => {
        expect(kitchen({recipe})).toMatchSnapshot();
    });
});
