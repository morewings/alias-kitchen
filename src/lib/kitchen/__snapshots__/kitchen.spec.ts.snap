// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`alias kitchen cooks babel configuration 1`] = `
{
  "@": "./src",
}
`;

exports[`alias kitchen cooks jest configuration 1`] = `
{
  "^@/(.*)$": "<rootDir>/src/$1",
}
`;

exports[`alias kitchen cooks rollup configuration 1`] = `
{
  "@": "/resolved/./src",
}
`;

exports[`alias kitchen cooks rspack configuration 1`] = `
{
  "@": "/resolved/./src",
}
`;

exports[`alias kitchen cooks vite configuration 1`] = `
{
  "@": "/resolved/./src",
}
`;

exports[`alias kitchen cooks webpack configuration 1`] = `
{
  "@": "/resolved/./src",
}
`;
